### ISSUE TITLE
Re-design the contracts architecture

### ISSUE THREAD

**Issue by bdim1** [supersaiyanxyz/supersaiyan-contracts#13]
Split the SuperSaiyanRegistry and ISuperSaiyanRegistry into multiple contracts. This is needed because of better modularity which improves readability and overall maintenance. Also we're reaching contract size limit and we need to implement additional features, which is an obvious sign for more modular approach.

**Comment by bdim1** [comment]
@trozler @zant This is a contract architecture I came up with, which I think is good for or project:

The main goal is to split the huge SuperSaiyanRegistry contract, which inherits from SuperAppBase, ISuperSaiyanRegistry and ERC1155. At the same time this contract is a SuperApp, NFT factory and service metadata manager. 
Defining the roles this contract performs:
- SuperApp - actions performed based on user subscriptions, leveraging the SuperFluid platform
- NFT factory - used for minting NFTs to users based on certain rules, leveraging the ERC1155 standard
- Service metadata manager - storage and management of the data and metadata for the entities in the SuperSaiyan ecosystem,

it is logical to split the SuperSaiyanRegistry contract in 3 separate contract:

#### SuperSaiyanApp
- inherits SuperAppBase (provided from SuperFluid)
- implements the `afterAgreementCreated`, `afterAgreementUpdated` and `afterAgreementTerminated` callbacks
- called when the users want to: open, close or update a subscription to specific service.
- calls SuperSaiyanRegistry and SuperSaiyanNftFactory for service and package requirements and for NFT minting logic
- the subscription requirement logic and nft minting logic based on nft templates is implemented in this contract

#### SuperSaiyanRegistry (new)
- inherits ISuperSaiyanRegistry (our own implementation)
- manages data and metadata related to all entities in the SuperSaiyan ecosystem
- the data is stored in this contract
- functions for managing services, packages and nft templates are part of this contract
- users interact with this contract to create services, add packages to services, add nft templates to services

#### SuperSaiyanNftFactory
- inherits ERC1155 (provided from OpenZeppelin)
- mints specific NFTs for users, and stores the balances (everything inherited from ERC1155)
- NFT minting only allowed from SuperSaiyanApp (only SuperSaiyanApp can mint NFTs for user, when the NFT template conditions are satisfied).

Please let me know your opinion on this, if you think it is not future compatible or there are shortcomings to this approach, please let me know. I do have a specific implementation in mind and can start doing that after the NFT template tests are finished.

**Comment by trozler** [comment]
Hi man nice work I think this makes sense. 

Breaking of the `SuperSaiyanNftFactory`  as described above makes complete sense. 
In regards to`SuperSaiyanApp` and `SuperSaiyanRegistry`.

Question 1:
In `SuperSaiyanApp` it says the contract is called when "called when the users want to: open, close or update a subscription to specific service. 

Whilst also in `SuperSaiyanRegistry` "users interact with this contract to create services, add packages to services, add nft templates to services"

Could you elaborate on how "open and closing of streams" is different than "creating service and add packages"?
Do you mean when you create service we will interact with `SuperSaiyanRegistry` which will then call `SuperSaiyanApp` to register stream?

**Comment by trozler** [comment]
@bdim1 Before we start the contract re-structure I think it will be important to think about the ENS integration. As this issue is quite critical from a UX perspective. 
See #15 for more information.

**Comment by bdim1** [comment]
> Hi man nice work I think this makes sense.
> 
> Breaking of the `SuperSaiyanNftFactory` as described above makes complete sense.
> In regards to`SuperSaiyanApp` and `SuperSaiyanRegistry`.
> 
> Question 1:
> In `SuperSaiyanApp` it says the contract is called when "called when the users want to: open, close or update a subscription to specific service.
> 
> Whilst also in `SuperSaiyanRegistry` "users interact with this contract to create services, add packages to services, add nft templates to services"
> 
> Could you elaborate on how "open and closing of streams" is different than "creating service and add packages"?
> Do you mean when you create service we will interact with `SuperSaiyanRegistry` which will then call `SuperSaiyanApp` to register stream?

Yes, opening and closing streams interacts with the superfluid framework in the background. Your reasoning about interaction is correct, except the registering a stream. We don't register a stream, a stream is started when the user "subscribes" to a certain service, in technical terms the user opens a flow to the superapp which is redirected to the service provider.

**Comment by trozler** [comment]
I see see thank you. 
Ok this makes sense to me. I'm happy with this architecture.
